src/security/kms_client.cpp

```diff
@@ -0,0 +1,182 @@
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <initializer_list>
+#include <map>
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+#include "absl/strings/escaping.h"
+#include "fmt/core.h"
+#include "http/http_client.h"
+#include "http/http_method.h"
+#include "http/http_status_code.h"
+#include "nlohmann/json.hpp"
+#include "nlohmann/json_fwd.hpp"
+#include "security/kms_client.h"
+#include "replica/replication_app_base.h"
+#include "utils/error_code.h"
+#include "utils/fmt_logging.h"
+
+namespace dsn {
+namespace security {
+
+dsn::error_s KMSClient::DecryptEncryptionKey(const dsn::replication::replica_kms_info &kms_info,
+                                             std::string *decrypted_key)
+{
+    nlohmann::json payload;
+    payload["name"] = cluster_key_name_;
+    std::string iv_plain = ::absl::HexStringToBytes(kms_info.iv);
+    std::string iv_b64;
+    ::absl::WebSafeBase64Escape(iv_plain, &iv_b64);
+    payload["iv"] = iv_b64;
+    std::string eek_plain = ::absl::HexStringToBytes(kms_info.eek);
+    std::string eek_b64;
+    ::absl::WebSafeBase64Escape(eek_plain, &eek_b64);
+    payload["material"] = eek_b64;
+
+    http_client client;
+    RETURN_NOT_OK(client.init());
+    RETURN_NOT_OK(client.set_auth(http_auth_type::SPNEGO));
+
+    std::vector<std::string> urls;
+    urls.reserve(kms_urls_.size());
+    for (const auto &url : kms_urls_) {
+        urls.emplace_back(fmt::format("{}/v1/keyversion/{}/_eek?eek_op=decrypt", url, kms_info.kv));
+    }
+    client.clear_header_fields();
+    client.set_content_type("application/json");
+    client.set_accept("*/*");
+
+    RETURN_NOT_OK(client.with_post_method(payload.dump()));
+
+    nlohmann::json j;
+    for (const auto &url : urls) {
+        RETURN_NOT_OK(client.set_url(url));
+        std::string resp;
+        auto err = client.exec_method(&resp);
+        if (err.code() == ERR_NETWORK_FAILURE || err.code() == ERR_TIMEOUT) {
+            continue;
+        } else {
+            RETURN_NOT_OK(err);
+        }
+        http_status_code http_status;
+        client.get_http_status(http_status);
+        if (http_status == http_status_code::kOk) {
+            try {
+                j = nlohmann::json::parse(resp);
+            } catch (nlohmann::json::exception &exp) {
+                LOG_ERROR("encode kms_info to json failed: {}, data = [{}]", exp.what(), resp);
+            }
+            break;
+        }
+        LOG_WARNING(
+            "The http status is ({}), and url is ({})", get_http_status_message(http_status), url);
+    }
+
+    std::string dek_b64;
+    if (j.contains("material")) {
```

Comment:
How about defining a new macro like `RETURN_ERR_S_NOT_TRUE` to simplify the code here and some other places.